docker/manager.py

```diff
@@ -0,0 +1,821 @@
+                    'major': major,
+                    'minor': minor,
+                    'shortened': ".".join([major, minor]),
+                },
+                cudnn_namespace: {
+                    'version': cudnn_release_version,
+                    'major_version': cudnn_major_version,
+                },
+                'components': self._cuda_components,
+            }
+            _build_ctx.pop('cuda_prefix_url')
+        else:
+            logging.debug("distro doesn't enable CUDA. Skipping...")
+            for key in list(_build_ctx.keys()):
+                if cuda_regex.match(key):
+                    _build_ctx.pop(key)
+
+        return _build_ctx
+
+    def render_dockerfile_from_templates(
+        self,
+        input_tmpl_path: pathlib.Path,
+        output_path: pathlib.Path,
+        ctx: Dict,
+        tags: Optional[Dict[str, str]] = None,
+    ):
+        with open(input_tmpl_path, 'r') as inf:
+            logging.debug(f"Processing template: {input_tmpl_path}")
+            name = input_tmpl_path.name
+            if DOCKERFILE_PREFIX not in name:
+                logging.debug(f"Processing {name}")
+                output_name = name[: -len('.j2')]
+            else:
+                output_name = DOCKERFILE_PREFIX.split('.')[1]
+
+            template = self._template_env.from_string(inf.read())
+            if not output_path.exists():
+                logging.debug(f"Creating {output_path}")
+                output_path.mkdir(parents=True, exist_ok=True)
+
+            with open(f"{output_path}/{output_name}", 'w') as ouf:
+                ouf.write(template.render(metadata=ctx, tags=tags))
+            ouf.close()
+        inf.close()
+
+    def generate_dockerfiles(self, target_dir, python_version):
+        """
+        Workflow:
+        walk all templates dir from given context
+
+        our targeted generate directory should look like:
+        generated
+            -- packages (model-server, yatai-service)
+                -- python_version (3.8,3.7,etc)
+                    -- {distro}{distro_version} (ubuntu20.04, ubuntu18.04, centos8, centos7, amazonlinux2, etc)
+                        -- runtime
+                            -- Dockerfile
+                        -- devel
+                            -- Dockerfile
+                        -- cudnn
+                            -- Dockerfile
+                        -- Dockerfile (for base deps images)
+
+        Args:
+            target_dir: parent directory of generated Dockerfile. Default: ./generated.
+            python_version: target list of supported python version. Overwrite default supports list if
+                            FLAGS.python_version is given.
+
+        Returns:
+            Let the magic happens.
+        """
+        _tag_metadata = defaultdict(list)
+
+        logging.info(f"Generating dockerfile to {target_dir}...\n")
+        for package in self._packages.keys():
```

Comment:
Feedback needed. As I mentioned above, I wrote this around 3am 😄 so there are like 5 nested for loops